geowebcache/s3storage/src/main/java/org/geowebcache/s3/TMSKeyBuilder.java

```diff
@@ -40,43 +41,43 @@
      * Key format, comprised of
      * {@code <prefix>/<layer name>/<gridset id>/<format id>/<parameters hash>/<z>/<x>/<y>.<extension>}
      */
-    private static final String TILE_FORMAT = "%s/%s/%s/%s/%s/%d/%d/%d.%s";
+    private static final String TILE_FORMAT = "%s%s/%s/%s/%s/%d/%d/%d.%s";
 
     /**
      * Coordinates prefix: {@code <prefix>/<layer name>/<gridset id>/<format id>/<parameters hash>/}
      */
-    private static final String COORDINATES_PREFIX_FORMAT = "%s/%s/%s/%s/%s/";
+    private static final String COORDINATES_PREFIX_FORMAT = "%s%s/%s/%s/%s/";
 
     /**
      * Layer prefix format, comprised of {@code <prefix>/<layer name>/}
      */
-    private static final String LAYER_PREFIX_FORMAT = "%s/%s/";
+    private static final String LAYER_PREFIX_FORMAT = "%s%s/";
 
     /**
      * layer + gridset prefix format, comprised of {@code <prefix>/<layer name>/<gridset id>/}
      */
-    private static final String GRIDSET_PREFIX_FORMAT = "%s/%s/%s/";
+    private static final String GRIDSET_PREFIX_FORMAT = "%s%s/%s/";
 
     public static final String LAYER_METADATA_OBJECT_NAME = "metadata.properties";
     public static final String PARAMETERS_METADATA_OBJECT_PREFIX = "parameters-";
     public static final String PARAMETERS_METADATA_OBJECT_NAME = 
             PARAMETERS_METADATA_OBJECT_PREFIX+"%s.properties";
 
-    private static final String STORE_METADATA_FORMAT = "%s/" + 
+    private static final String STORE_METADATA_FORMAT = "%s" +
             LAYER_METADATA_OBJECT_NAME;
     private static final String LAYER_METADATA_FORMAT = "%s/%s/" + 
             LAYER_METADATA_OBJECT_NAME;
-    private static final String PARAMETERS_METADATA_FORMAT = "%s/%s/" + 
+    private static final String PARAMETERS_METADATA_FORMAT = "%s%s/" + 
             PARAMETERS_METADATA_OBJECT_NAME;
     private static final String PARAMETERS_METADATA_PREFIX_FORMAT = 
-            "%s/%s/" + PARAMETERS_METADATA_OBJECT_PREFIX;
+            "%s%s/" + PARAMETERS_METADATA_OBJECT_PREFIX;
 
     private String prefix;
 
     private TileLayerDispatcher layers;
 
     public TMSKeyBuilder(final String prefix, TileLayerDispatcher layers) {
-        this.prefix = prefix;
+        this.prefix = Strings.isNullOrEmpty(prefix) ? "" : prefix + "/";
```

Comment:
I really dislike how this confuses things.  Most of the `/`s are in the templates, but then this one special slash is being added here.  This is a really good opportunity for future mistakes.

Maybe this is a case for changing from using `String.format()` with templates that all do essentially the same thing for varying numbers of parameters to say just using `String.join()`.

Reply:
@smithkm I didn't like it much either, but the only straight-forward alternatives I saw was to add an IF structure to every single method with either double the amount of templates or each time removing the first character. So I thought this was the least evil.

But I like your suggestion of String.join, I will change this PR.

Reply:
The String.join method doesn't ignore nulls, in fact it will add the word "null" to it and still include an extra delimiter. Will have to write our own method.

Reply:
```java
Arrays.stream(elements).filter(Objects::nonNull).collect(Collectors.joining("/"))
```

Reply:
Come to think of it, it should probably be escaping in case of elements containing slashes, but it's too late to add that as it would break backward compatibility for anyone who already has a layer or whatever with a slash in the name.